### PR TITLE
chore: validate submission accounts on deserialization

### LINKED ISSUE #4449: chore: validate submission accounts on deserialization
### Background

Original comment: https://github.com/cowprotocol/services/pull/4430#discussion_r3310124860

The validation can be made part of the deserialization process by using a wrapper struct that ensures the values inside always match the required invariant.

### Details

Something like as follows would work:

```
struct SubmissionAccounts(Vec<Account>);

impl SubmissionAccounts {
    fn new(accounts: Vec<Accounts>) -> Result<Self, Err> {
         // validate that all accounts match the invariant
    }
}

// Deserialization implementation
```

### Acceptance criteria
* New struct
* Custom deserialization
* Updated tests — i.e. use the constructor to + unwrap/expect to ensure the values are correct

### PR DESCRIPTION
**Pull request by dymchenkko**
# Description

`submission-accounts` is already validated at runtime in `solver::Config::validate()` (added in #4430). As suggested in [this review comment](https://github.com/cowprotocol/services/pull/4430#discussion_r3310124860), this additionally enforces the "submission accounts must be signers" invariant at config **deserialization** via a wrapper type, so an invalid config fails to load and the type can't be constructed in an invalid state.

# Changes

* Add a `SubmissionAccounts(Vec<Account>)` wrapper in `config/file/mod.rs` whose `new()` constructor and custom `Deserialize` impl reject any address-only `Account` (only signers can sign EIP-7702 delegated settlements). Config loading now fails fast with a clear error.
* Use `SubmissionAccounts` for the `submission_accounts` config field; `load.rs` unwraps the validated list via `into_inner()`.
* The existing runtime check in `solver::Config::validate()` is left in place as defense-in-depth (also guards programmatic construction); no change there.
* Add TOML deserialization tests: omitted/default → empty, signer accounts accepted, and address-only accounts rejected during deserialization.

## How to test

1. `cargo test -p driver --lib submission_accounts`
2. Manually: add an address-only entry to `submission-accounts` in a driver config (e.g. `submission-accounts = ["0x0000000000000000000000000000000000000002"]`) and start the driver — it refuses to start with `EIP-7702 submission accounts must be signers; address-only accounts cannot sign delegated settlement transactions`. A private-key entry starts normally.

## Related Issues

Fixes #4449

### CODE REVIEW

File: crates/driver/src/infra/config/file/mod.rs [hunk 1]
```diff
@@ -365,6 +365,52 @@ enum Account {
...
 
+/// Accounts used for parallel EIP-7702 settlement submission. Every account is
+/// guaranteed to be a signer (never [`Account::Address`]), because address-only
+/// accounts cannot sign the delegated settlement transactions. The invariant is
+/// enforced on deserialization, so any constructed value is valid by
+/// construction.
+#[derive(Debug, Default)]
+struct SubmissionAccounts(Vec<Account>);
+
+impl SubmissionAccounts {
```
Comment (milaforge) [inline crates/driver/src/infra/config/file/mod.rs:374]: Nice work!

One gap: the current tests still cover the runtime validation path, but not TOML deserialization itself. I’d add focused tests for:

1. default/empty accounts, 
2. signer accounts accepted, 
3. and address-only accounts rejected during deserialization.

File: crates/driver/src/infra/config/file/mod.rs [hunk 2]
```diff
@@ -365,6 +365,52 @@ enum Account {
...
+impl<'de> Deserialize<'de> for SubmissionAccounts {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let accounts = Vec::<Account>::deserialize(deserializer)?;
+        Self::new(accounts).map_err(serde::de::Error::custom)
+    }
+}
+
```
Comment (milaforge) [inline crates/driver/src/infra/config/file/mod.rs:414]: Could we add focused tests here for omitted/default accounts, signer accounts accepted, and address-only accounts rejected during TOML deserialization?

**Comment by milaforge** [comment]
@dymchenkko Small PR-body nit: it says runtime validation was removed and deserialization tests were added, but the final diff keeps the validation and does not yet add those tests.



https://github.com/cowprotocol/services/blob/8b0b9f46208b69172bfe262a25be4245ee2f6f6d/crates/driver/src/infra/solver/mod.rs#L239

https://github.com/cowprotocol/services/blob/8b0b9f46208b69172bfe262a25be4245ee2f6f6d/crates/driver/src/infra/solver/eip7702.rs#L93

**Comment by dymchenkko** [comment]
@failuresmith Thanks for the review! Applied the changes.

**Comment by milaforge** [comment]
Longer term, I think the wrapper may belong in `infra::solver`, not the file-config layer. The invariant is about runtime solver behavior, so ideally `solver::Config` would store `SubmissionAccounts` instead of `Vec<Account>`, and both TOML loading and programmatic construction would go through the same constructor.

The current PR still improves the file-loading path and keeps `validate()` as a fallback, so I don’t think it has to block this change. But a domain-level wrapper would avoid maintaining the same invariant in two places.

**Comment by milaforge** [comment]
If we want **one canonical check**, it should live **just in the runtime/domain type**, not only in file config.

The stronger design is:

```rust
// infra::solver
pub struct SubmissionAccounts(Vec<Account>);

impl SubmissionAccounts {
    pub fn new(accounts: Vec<Account>) -> anyhow::Result<Self> {
        // reject Account::Address once
    }
}
```

Then:

```rust
pub struct Config {
    pub submission_accounts: SubmissionAccounts,
}
```

Now both paths must go through the same constructor:

- TOML loading parses `file::Account`, converts to `solver::Account`, then calls `solver::SubmissionAccounts::new(...)`.
- Programmatic construction also has to call `solver::SubmissionAccounts::new(...)`.

---

The tradeoff is scope: moving it into `solver::Config` touches more call sites than #4449 likely intended.

**Comment by dymchenkko** [comment]
Yeah, I was thinking about this too,  a `solver::SubmissionAccounts` would be a nice tightening. I'd lean toward keeping it as a follow-up though: the issue title is specifically "validate on deserialization," and the current design fails fast at parse time, before any (potentially async) KMS signer loading. Moving the wrapper into `solver::Config` also expands scope beyond what #4449 targets. @jmg-duarte what do you think? I can open a follow-up issue for the domain-level refactor if needed.

File: crates/driver/src/infra/config/file/mod.rs [hunk 3]
```diff
@@ -365,6 +365,52 @@ enum Account {
...
+/// Accounts used for parallel EIP-7702 settlement submission. Every account is
+/// guaranteed to be a signer (never [`Account::Address`]), because address-only
+/// accounts cannot sign the delegated settlement transactions. The invariant is
+/// enforced on deserialization, so any constructed value is valid by
+/// construction.
+#[derive(Debug, Default)]
+struct SubmissionAccounts(Vec<Account>);
+
+impl SubmissionAccounts {
+    fn new(accounts: Vec<Account>) -> Result<Self, InvalidSubmissionAccounts> {
```
Comment (jmg-duarte) [inline crates/driver/src/infra/config/file/mod.rs:378]: Document the check/invariant + add two tests pass/fail for this

File: crates/driver/src/infra/config/file/mod.rs [hunk 4]
```diff
@@ -365,6 +365,52 @@ enum Account {
     Address(eth::Address),
 }
 
+/// Accounts used for parallel EIP-7702 settlement submission. Every account is
+/// guaranteed to be a signer (never [`Account::Address`]), because address-only
+/// accounts cannot sign the delegated settlement transactions. The invariant is
+/// enforced on deserialization, so any constructed value is valid by
+/// construction.
```
Comment (jmg-duarte) [inline crates/driver/src/infra/config/file/mod.rs]:
```suggestion
/// accounts cannot sign the delegated settlement transactions.
```
Not relevant here

File: crates/driver/src/infra/config/file/mod.rs [hunk 5]
```diff
@@ -1082,4 +1128,46 @@ mod tests {
...
+    }
+
+    #[test]
+    fn submission_accounts_accepts_signers() {
+        let config: SubmissionAccountsConfig = toml::from_str(
+            r#"
+            submission-accounts = [
+                "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d",
+            ]
+        "#,
```
Comment (jmg-duarte) [inline crates/driver/src/infra/config/file/mod.rs]: can we match the indentation on start and end? #" and "#

**Review by jmg-duarte — CHANGES_REQUESTED** [review]
Let's not expand the scope. You can open a follow up draft for evaluation (no guarantees on whether it will be considered useful), this was the most crucial piece

**Comment by dymchenkko** [comment]
@jmg-duarte just applied changes: trimmed the wrapper doc; added doc on `new()` + two direct tests (`..._new_accepts_signers` / `…_new_rejects_address_only`) and aligned raw-string delimiters in the TOML tests.

**Review by jmg-duarte — APPROVED** [review]
